gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java

```diff
@@ -689,83 +702,339 @@ public void write(JsonWriter out, Currency value) throws IOException {
       }.nullSafe();
   public static final TypeAdapterFactory CURRENCY_FACTORY = newFactory(Currency.class, CURRENCY);
 
+  /**
+   * An abstract {@link TypeAdapter} for classes whose JSON serialization consists of a fixed set of
+   * integer fields. That is the case for the legacy serialization of various {@code java.time}
+   * types.
+   *
+   * <p>This is a base class for {@link Calendar}, {@link Duration}, {@link Instant}, {@link
+   * LocalDate}, {@link LocalTime}, {@link LocalDateTime}, and {@link ZoneOffset}.
+   */
+  private abstract static class IntegerFieldsTypeAdapter<T> extends TypeAdapter<T> {
+    private final List<String> fields;
+
+    IntegerFieldsTypeAdapter(String... fields) {
+      this.fields = Arrays.asList(fields);
+    }
+
+    abstract T create(long[] values);
+
+    abstract long[] integerValues(T t);
+
+    @Override
+    public T read(JsonReader in) throws IOException {
+      if (in.peek() == JsonToken.NULL) {
```

Comment:
Not any sort of blocker for this PR, but: Has there been discussion of a common `TypeAdapter` subclass (even if only a private one for our own use) that passes through nulls, leaving subclasses free to implement `readNonNull` and `writeNonNull` methods without the null-handling ceremony? Maybe the common class could even call `beginObject` and `endObject` (if it's not dangerous that that delays the call to `endObject` until after the user subclass has done its thing)?